### PR TITLE
add e2e for h2c

### PR DESCRIPTION
**Pull request by itsmurugappan**
## Description

Add e2e test for creating a grpc service with port name as h2c

## Changes

* use the grpc ping service from knative/serving repo
* create service with custom port details and validate

### CODE REVIEW

File: test/upload-test-images.sh [hunk 1]
```diff
@@ -32,6 +33,9 @@ function upload_test_images() {
   # so the resulting yaml produced is ignored.
   # We limit the number of concurrent builds (jobs) to avoid OOMs.
   ko resolve --jobs=4 ${tag_option} -RBf "${image_dir}" > /dev/null
+
+  #build and publish images from vendor directory
+  ko resolve --jobs=4 ${tag_option} -RBf "${vendor_image_dir}" > /dev/null
```
Comment (navidshaikh) [inline test/upload-test-images.sh:38]: why do we need this?

Comment (itsmurugappan) [reply test/upload-test-images.sh:38]: Instead of creating a new service. Building the grpc server image from knative serving test repo.

Comment (navidshaikh) [reply test/upload-test-images.sh:38]: there are other test images as well in serving, but it only picks up the grpc-ping, how is that handled ? is that `pkgtest.ImagePath("grpc-ping")` ?
```
2020/09/09 05:03:15 Using base gcr.io/distroless/static:nonroot for knative.dev/serving/test/test_images/grpc-ping
2020/09/09 05:03:16 Building knative.dev/serving/test/test_images/grpc-ping
2020/09/09 05:04:11 Publishing gcr.io/knative-boskos-51/kclient-e2e-img/8803/grpc-ping:latest
2020/09/09 05:04:13 Published gcr.io/knative-boskos-51/kclient-e2e-img/8803/grpc-ping@sha256:f8f53534f1a74fd7e9e4df8338c55ee8c8fee29fd8aac2f03cbacdaacbdbce72
=============================
```

also, seeing a lot of vendored files, I think thats required for test images to be built? Can we avoid vendoring them and download them part of image build in test execution ?

Comment (itsmurugappan) [reply test/upload-test-images.sh:38]: added "knative.dev/serving/test/test_images/grpc-ping" in hack/tools.go to bring in this service, which is bringing in other dependencies to the vendor.

File: test/upload-test-images.sh [hunk 2]
```diff
@@ -22,6 +22,7 @@ function upload_test_images() {
   # to pickup .ko.yaml
   cd "$( dirname "$0")/.."
   local image_dir="test/test_images"
+  local vendor_image_dir="vendor/knative.dev/serving/test/test_images"
```
Comment (dsimansk) [inline test/upload-test-images.sh:25]: @itsmurugappan @navidshaikh 
I was wondering how you would feel about creating a symlink of grpc-ping from vendor dir to `./test/test-images/grpc-ping`. It'd make test dependencies appear a bit more unified. In addition there'd be a benefit for our internal repo handling. But it's just an idea how to make our life easier elsewhere. :)

File: test/test_images/grpc-ping [hunk 3]
```diff
@@ -0,0 +1 @@
+/Developer/work/go/src/knative.dev/client/vendor/knative.dev/serving/test/test_images/grpc-ping
```
Comment (dsimansk) [inline test/test_images/grpc-ping:1]: It should be relative link, something like the following should do the trick.
```
ln -s -r vendor/knative.dev/serving/test/test_images/grpc-ping test/test_images/grpc-ping
```

Comment (itsmurugappan) [reply test/test_images/grpc-ping:1]: i dont see a -r option in mac.. not sure what am missing

Comment (dsimansk) [reply test/test_images/grpc-ping:1]: I guess older `ln` version vs. gnu `ln`. For me It works also relatively to link without `-r` option.

```
cd test/test_images
ln -s ../../vendor/knative.dev/serving/test/test_images/grpc-ping grpc-ping
```

File: lib/test/service.go [hunk 4]
```diff
@@ -122,3 +122,15 @@ func ValidateServiceResources(r *KnRunResultCollector, serviceName string, reque
...
+//GetServiceFromKNServiceDescribe runs the kn service describe command
+//decodes it into a ksvc and returns it.
+func GetServiceFromKNServiceDescribe(r *KnRunResultCollector, serviceName string) servingv1.Service {
+	out := r.KnTest().Kn().Run("service", "describe", serviceName, "-ojson")
+	data := json.NewDecoder(strings.NewReader(out.Stdout))
+	data.UseNumber()
+	var service servingv1.Service
+	err := data.Decode(&service)
+	assert.NilError(r.T(), err)
+	return service
```
Comment (navidshaikh) [inline lib/test/service.go:135]: we can return here the pointer to service object instead

**Review by navidshaikh — APPROVED** [review]
/lgtm
/approve
/hold
/assign @dsimansk 

David, unhold if you approve.

File: lib/test/service.go (hunk 4 above)
Comment (navidshaikh) [reply lib/test/service.go:135]: (can be fixed in a subsequent PR as well)

Comment (dsimansk) [reply lib/test/service.go:135]: Nice catch. Let's get this in and address in follow-up PR then.

**Comment by dsimansk** [comment]
@itsmurugappan thanks for bearing with me and symlinks. :)

Looks good to me, let's get this in and address Navid's comment later.
/unhold

**Pull request merged**